### PR TITLE
esp_lvgl_port: add sw_rotate and psram to sram. (BSP-417)

### PR DESCRIPTION
**Pull request by espressif2022**
# ESP-BSP Pull Request checklist

> Note: For new BSPs create a PR with this [link](https://github.com/espressif/esp-bsp/compare/main...my-branch?quick_pull=1&template=pr_template_bsp.md).

- [ ] Version of modified component bumped
- [ ] CI passing

# Change description
1: add sw rotate.
2: add sram copy form PSRAM.

### CODE REVIEW

**Review by espzav — COMMENTED** [review]
@espressif2022 Thank you very much for this PR. LGTM. It will be nice to add some note into readme for this feature. For example on the end of the file (after rotating screen). There can be chapter - "Using PSRAM" or something like that. And describe, how to use it in example and why we need SW rotation.

File: components/esp_lvgl_port/include/esp_lvgl_port.h [hunk 1]
```diff
@@ -48,6 +48,7 @@ typedef struct {
  * @brief Rotation configuration
  */
 typedef struct {
+    bool sw_rotate;/*!< Use software rotation (slower) */
```
Comment (espzav) [inline components/esp_lvgl_port/include/esp_lvgl_port.h:51]: I think, that better place for `bool sw_rotate;` is inside flags, in main structure (`lvgl_port_display_cfg_t`).

File: components/esp_lvgl_port/include/esp_lvgl_port.h [hunk 2]
```diff
@@ -61,6 +62,7 @@ typedef struct {
     esp_lcd_panel_handle_t panel_handle;    /*!< LCD panel handle */
     uint32_t    buffer_size;    /*!< Size of the buffer for the screen in pixels */
     bool        double_buffer;  /*!< True, if should be allocated two buffers */
+    uint32_t    trans_size;     /*!< Allocated buffer will be in SRAM to move framebuf */
```
Comment (espzav) [inline components/esp_lvgl_port/include/esp_lvgl_port.h:65]: I think, better name for it is: `sw_rotation_buffer` or somethink like that.

**Review by espzav — COMMENTED** [review]
@espressif2022 Thank you for this PR. I tested in on HW, but I don't see any difference in all cases. 
I left some comments. 
@tore-espressif PTAL

File: components/esp_lvgl_port/include/esp_lvgl_port.h (hunk 2 above)
Comment (espressif2022) [reply components/esp_lvgl_port/include/esp_lvgl_port.h:65]: >Why is it named trans_size, not sw_rotation_buffer?

- The operation of sw_rotate is inside LVGL, and I originally planned to do sw_rotate in lvgl_port_flush_callback, but considering that LVGL supports sw_rotation inside, so I've deleted this part.

- The function of this buffer here is that when the frame_buffer is in the PSRAM, it provides a small buffer of internal SRAM to deal with it, so SPI driver no needs to do malloc and free for each transfer internally. It is user-configurable and more flexible.

>I tested in on HW, but I don't see any difference in all cases.

The phenomenon is like the screenshot I gave you, there will be an oblique tear, which is more obvious on the customer's product. 
Let me think how to show it. I tried it on box, but I can't show it well.

Comment (espressif2022) [reply components/esp_lvgl_port/include/esp_lvgl_port.h:65]: When I'm testing the demo, I found some compatibility issues in lvgl 8.3.9 and I'm confirming.

Comment (espressif2022) [reply components/esp_lvgl_port/include/esp_lvgl_port.h:65]: This is a 240*320 resolution screen, and customers need to rotate 90 degrees to use it horizontally.
[LCD.png](https://cowtransfer.com/s/c2b2b50664d940)

When using hardware rotation, a more noticeable oblique tear can be seen.
[ hw_roate.mp4 ](https://cowtransfer.com/s/185c869d77704b) 
[ sw_rotate.mp4 ](https://cowtransfer.com/s/4b3b65af8df848)

This is a comparison of the boot video, using full-screen framebuf + 10-line SRAM handling and 10-line SRAM rendering, and the tearing of the interface is also significantly improved.
[ poweron.mp4 ](https://cowtransfer.com/s/7cad37fcd7e748)

File: components/esp_lvgl_port/include/esp_lvgl_port.h [hunk 3]
```diff
@@ -61,9 +61,11 @@ typedef struct {
...
     uint32_t    hres;           /*!< LCD display horizontal resolution */
     uint32_t    vres;           /*!< LCD display vertical resolution */
     bool        monochrome;     /*!< True, if display is monochrome and using 1bit for 1px */
+    bool        sw_rotate;      /*!< Use software rotation (slower) */
```
[the remark this answers is not in the document]
Comment (espressif2022) [reply components/esp_lvgl_port/include/esp_lvgl_port.h:68]: sorry, I'll update and check it tomorrow morning.

**Review by espzav — APPROVED** [review]
@espressif2022 Thank you for this PR. Tested on HW again. LGTM
Please, don't forget squash commits. And waiting for @tore-espressif review.

File: components/esp_lvgl_port/README.md [hunk 4]
```diff
@@ -182,11 +182,37 @@ Every LVGL calls must be protected with these lock/unlock commands:
...
+
+If the SRAM is insufficient, you can use the PSRAM as a canvas and use a small trans_buffer to carry it, this makes drawing more efficient.
+``` c
+    .buffer_size = DISP_WIDTH*DISP_HEIGHT, // in PSRAM, not DMA-capable
+    .trans_size = size, // in SRAM, DMA-capable
+    .flags = {
+        .buff_spiram = true,
+        .buff_dma = false,
+    }
+```
```
[the remark this answers is not in the document]
Comment (tore-espressif) [reply components/esp_lvgl_port/README.md:219]: +1

File: components/esp_lvgl_port/include/esp_lvgl_port.h (hunk 2 above)
Comment (tore-espressif) [reply components/esp_lvgl_port/include/esp_lvgl_port.h:64]: > better name for it is: sw_rotation_buffer or something like that

i'd also like to see more descriptive name for this struct member

**Review by tore-espressif — COMMENTED** [review]
@espressif2022 I'm sorry for the very late review.

Could you please describe the role of `disp_ctx->trans_sem` ?

Otherwise LGTM

**Comment by espressif2022** [comment]
> @espressif2022 I'm sorry for the very late review.
> 
> Could you please describe the role of `disp_ctx->trans_sem` ?
> 
> Otherwise LGTM

Hello, later for this.
This is to notify `lvgl_port_flush_callback` to continue sending the next segment of data when the current frame is finished(`lvgl_port_flush_ready_callback`)
It's a bit like the bounce_buffer mechanism like RGB driver, which allocates the framebuffer to the PSRAM, and then assigns a small internal buffer to carry it and transmit it in segments.

**Pull request merged**